projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs

```diff
@@ -283,28 +279,51 @@ public void Write(OutgoingFrame frame)
 
         private async Task ReadLoop()
         {
-            while (!FrameReader.Completion.IsCompleted)
+            bool allowSync = true;
+            try
             {
-                // Let's read some bytes
-                if (!_pipeReader.TryRead(out ReadResult readResult))
+                while (true)
                 {
-                    readResult = await _pipeReader.ReadAsync().ConfigureAwait(false);
-                }
+                    // Let's read some bytes
+                    if (!(allowSync && _pipeReader.TryRead(out ReadResult readResult)))
+                    {
+                        readResult = await _pipeReader.ReadAsync().ConfigureAwait(false);
+                    }
 
-                ReadOnlySequence<byte> buffer = readResult.Buffer;
-                if (buffer.IsEmpty)
-                {
-                    _frameWriter.Complete(new EndOfStreamException());
-                    return;
-                }
+                    int handled = 0;
+                    ReadOnlySequence<byte> buffer = readResult.Buffer;
+                    if (!buffer.IsEmpty)
```

Comment:
this is redundand

Reply:
It might still return 0 even if the buffer is not empty, since we might not have an entire frame to read.